### PR TITLE
Fix CUDA illegal memory access in embedding_bag per_sample_weights backward

### LINKED ISSUE #192445: [CUDA] illegal memory read on _embedding_bag_per_sample_weights_backward_kernel
### 🐛 Describe the bug

```python
import torch
import torch.nn.functional as F


torch.cuda.init()
gpu = torch.device("cuda")
weight = torch.ones((5, 2), dtype=torch.float16, device=gpu)
indices = torch.zeros((1024,), dtype=torch.int32, device=gpu)
indices_alias = torch.from_dlpack(indices)
offsets = torch.tensor([0], dtype=torch.int32, device=gpu)
per_sample_weights = torch.ones(
    (1024,), dtype=torch.float16, device=gpu, requires_grad=True
)

output = F.embedding_bag(
    indices,
    weight,
    offsets,
    mode="sum",
    per_sample_weights=per_sample_weights,
)
indices_alias.fill_(2147483647)
output.sum().backward()
torch.cuda.synchronize()
```

```
========= COMPUTE-SANITIZER
========= Invalid __global__ read of size 2 bytes
=========     at void at::native::_embedding_bag_per_sample_weights_backward_kernel<c10::Half, int>(const T1 *, long, long, const T1 *, long, long, const T2 *, const T2 *, long, long, T1 *, T2)+0x520
=========     by thread (96,0,0) in block (0,0,0)
=========     Address 0x7f2291dffffc is out of bounds
=========     and is 8443133437 bytes after the nearest allocation at 0x7f209aa00000 of size 512 bytes
```

### Versions

/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  104
On-line CPU(s) list:                     0-103
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Gold 5320 CPU @ 2.20GHz
CPU family:                              6
Model:                                   106
Thread(s) per core:                      2
Core(s) per socket:                      26
Socket(s):                               2
Stepping:                                6
CPU(s) scaling MHz:                      39%
CPU max MHz:                             3400.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4400.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               2.4 MiB (52 instances)
L1i cache:                               1.6 MiB (52 instances)
L2 cache:                                65 MiB (52 instances)
L3 cache:                                78 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-25,52-77
NUMA node1 CPU(s):                       26-51,78-103
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.29.3
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.13.0+cu126
[pip3] triton==3.7.1
[conda] Could not collect

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia @malfet

### PR DESCRIPTION
**Pull request by PannenetsF**
Fixes #192445.

## Root cause

`_embedding_bag_per_sample_weights_backward_kernel` in
`aten/src/ATen/native/cuda/EmbeddingBag.cu` indexed the embedding `weight` by
`embedding_idx = indices[sample_idx]` with no bounds check, unlike the forward
`EmbeddingBag` kernels in the same file which validate the index against
`numRows`. When an index is out of range, the kernel reads past the `weight`
allocation, causing an illegal global memory access (undefined behavior /
potential silent corruption). This can be triggered when `indices` is mutated in
place after the forward pass (e.g. via a DLPack alias), so the bad value is only
seen at backward time.

A secondary defect was the lossy `const int embedding_idx = (int)indices[...]`
cast: for int64 indices this truncates to 32 bits, so a large in-range int64
index could alias the wrong row or defeat a naive check.

## Fix

Thread `weight.size(0)` into the kernel as `num_embeddings`, read the index with
its native `index_t` type (no truncation), and add a `CUDA_KERNEL_ASSERT` bounds
check against `[0, num_embeddings)` before the weight read, using the same
message convention as the forward kernels. This converts the illegal access into
a deterministic device-side assert.

## Impact

The change is confined to the CUDA per-sample-weights backward path. In-range
indices behave exactly as before (the assert is a no-op on the valid range); the
only behavioral change is that a previously-illegal out-of-range read now fails
fast with a device-side assert instead of reading out of bounds.

## Validation

New device-generic test `test_embedding_bag_per_sample_weights_mutated_indices`
reproduces the issue out-of-process (a device-side assert corrupts the CUDA
context) by mutating `indices` via a DLPack alias after the forward pass, then
running backward, and asserts the device-side assert markers appear in stderr.
It is gated `@onlyOn(["cuda"])` for both int32 and int64.

```
python test/nn/test_embedding.py -k test_embedding_bag_per_sample_weights_mutated_indices -v
```

Validation limitation: the CUDA test and `lintrunner -a` could not be run in the
authoring environment (macOS, no CUDA/GPU, PyTorch not built, lintrunner/spin
unavailable). These need to be run on a CUDA box; CI is expected to exercise them
here.

---

This pull request was prepared with the assistance of an AI coding assistant. The
author has reviewed and understands the code change and takes responsibility for
it.

### CODE REVIEW

**Comment by eqy** [comment]
@claude review this PR, does it introduce additional performance overheads from converting previous 32-bit integer arithmetic to 64-bit, and was the original issue a realistic one or an adversarial example discovered from a fuzzer?

**Review by eqy — CHANGES_REQUESTED** [review]
fix lint failures

**Comment by PannenetsF** [comment]
@eqy Thanks for the review. I've pushed an update addressing the feedback:
- **Lint** fixed (`lintrunner -a` clean).
- **`offset2bag` now bounds-checked** at native width (removed the lossy `int` cast on both indices; added `CUDA_KERNEL_ASSERT` against `num_bags`/`num_embeddings`, threaded in at the launch site).
- **Test now matches the specific assertion strings** (`embedding_idx >= 0 && embedding_idx < num_embeddings` / `bag_idx >= 0 && bag_idx < num_bags`) instead of the generic message, keeping the ROCm launch-failure fallbacks.
- **Subprocess harness** replaced with plain `subprocess.run(..., capture_output=True, text=True)` inspecting `stdout + stderr`.
- Added a **large-index truncation regression test** (`@largeTensorTest`) and coverage of the **negative-index** branch.

On your two questions: no meaningful perf cost (the address math was already 64-bit; the change removes one truncation and adds a per-sample compare outside the inner loop), and the original repro is adversarial, but the truncation half fixes real silent corruption for >2^31-row tables. PTAL.

**Review by eqy — APPROVED** [review]

**Comment by PannenetsF** [comment]
Verified this fix on an NVIDIA H20 (sm90) GPU: built the branch from source
(CUDA 13.1) and ran the two new tests. All 9 variants pass.

```
$ PYTORCH_TEST_WITH_SLOW=1 python -m pytest test/nn/test_embedding.py -v \
    -k "(test_embedding_bag_per_sample_weights_rejects_invalid_saved_indices \
         or test_embedding_bag_per_sample_weights_large_index) and cuda"

test_embedding_bag_per_sample_weights_large_index_cuda PASSED
test_..._rejects_invalid_saved_indices_target_indices_bound_negative_cuda_int32 PASSED
test_..._rejects_invalid_saved_indices_target_indices_bound_negative_cuda_int64 PASSED
test_..._rejects_invalid_saved_indices_target_indices_bound_upper_cuda_int32   PASSED
test_..._rejects_invalid_saved_indices_target_indices_bound_upper_cuda_int64   PASSED
test_..._rejects_invalid_saved_indices_target_offset2bag_bound_negative_cuda_int32 PASSED
test_..._rejects_invalid_saved_indices_target_offset2bag_bound_negative_cuda_int64 PASSED
test_..._rejects_invalid_saved_indices_target_offset2bag_bound_upper_cuda_int32 PASSED
test_..._rejects_invalid_saved_indices_target_offset2bag_bound_upper_cuda_int64 PASSED

9 passed, 344 deselected in 23.60s
```

- `test_embedding_bag_per_sample_weights_large_index` confirms the `> 2**31`
  index path now produces the correct gradient (previously the `(int)` cast
  truncated the index).
- `test_embedding_bag_per_sample_weights_rejects_invalid_saved_indices`
  (indices/offset2bag x negative/upper x int32/int64) confirms the new
  `CUDA_KERNEL_ASSERT` bounds checks fire as expected.

This verification was performed with the help of an AI assistant.
